// fs-filesystem/lib/utilities.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.emptyVolume = exports.emptyDevice = exports.splitEOL = exports.getYesNo = exports.hasSubstr = exports.stringify = exports.lasti = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _os = require('os');

var _os2 = _interopRequireDefault(_os);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// --------------------------------------
// Common Core - Utility Functions
// --------------------------------------

var lasti = exports.lasti = function lasti(a) {
  return a[a.length - 1];
};
var stringify = exports.stringify = function stringify(v) {
  return v && ((typeof v === 'undefined' ? 'undefined' : _typeof(v)) === 'object' || (typeof v === 'undefined' ? 'undefined' : _typeof(v)) === 'symbol') ? v.toString() : '' + v;
};
var hasSubstr = exports.hasSubstr = function hasSubstr(s, sub) {
  return s.indexOf(sub) !== -1;
};
var getYesNo = exports.getYesNo = function getYesNo(v) {
  return v === 'Yes' ? true : v === 'No' ? false : undefined;
};
var splitEOL = exports.splitEOL = function splitEOL(s) {
  return typeof s === 'number' ? function (v) {
    return v.split(_os2.default.EOL.repeat(s));
  } : s.split(_os2.default.EOL);
};

// --------------------------------------
// Common Core - Empty FS Objects
// --------------------------------------

var emptyDevice = exports.emptyDevice = function emptyDevice() {
  return {
    id: null,
    node: null,
    whole: false,
    parent: null,
    name: null,
    size: null,
    description: null,
    protocol: null,
    blockSize: null,
    readOnly: null,
    removable: null
  };
};

var emptyVolume = exports.emptyVolume = function emptyVolume() {
  return {
    id: null,
    node: null,
    whole: false,
    parent: null,
    name: null,
    description: null,
    blockSize: null,
    blocks: null,
    readOnly: null,
    mounted: null,
    mountPoint: null,
    partitionType: null,
    fs: null,
    space: { total: null, available: null, used: null }
  };
};
